### PR TITLE
Refactor multipart decoding implementation and add new API

### PR DESCRIPTION
**Pull request by yawkat**
This PR adds a new lower-level post body parsing API (PostBodyDecoder). Right now the only implementation is the MultipartDecoder because that's the difficult one, but I plan to implement URL decoding too.

The new API is purely stream-based and leaves the data management (e.g. writing to disk) to the caller. This makes it more compatible with netty 5 buffers, makes buffer lifecycles easier to understand to avoid buffer leaks, and simplifies the implementation to allow for future optimization.

Because multipart is incredibly complex and the old decoder is the result of years of issue reports for compatibility with obscure clients, and I don't have confidence in the existing test coverage, my main focus in this PR has been validating compatibility. To do this, I've ported HttpPostMultipartRequestDecoder to use the new low-level API internally, and created a fuzz test (MultipartComparisonTest) that compares the output with that of the old decoder (HttpPostMultipartRequestDecoderLegacy). This gives me high confidence that there is no feature regression.

In this process, I've found a few behaviors of the old decoder that I consider bugs not worth keeping around. Nonetheless, I've decided to keep HttpPostMultipartRequestDecoder behaving the same way (also to make the fuzzing work), by introducing a "quirk mode" for the streaming decoder. This quirk mode reproduces the small oddities of the original decoder during parsing.

This PR is against netty 5, but if it goes well, I plan to backport the streaming parser to netty 4. My suggestion is to keep the backport in netty-contrib, release a netty 4 version of codec-multipart, and eventually deprecate the form decoder in the netty 4 core repo.

### CODE REVIEW

**Comment by yawkat** [comment]
@violetagg 

@normanmaurer @trustin 

WDYT?

**Comment by yawkat** [comment]
I've added a `application/x-www-form-urlencoded` implementation, and migrated the old HttpPostStandardRequestDecoder to it. That means the new decoder API can now act as a full replacement for the InterfaceHttpPostRequestDecoders.

**Comment by altro3** [comment]
@violetagg @normanmaurer @trustin Could you see to this?

**Comment by loicmathieu** [comment]
Hi,
This PR fixes #12729 which we are affected by.
Any progress on it?

**Comment by anna-geller** [comment]
@violetagg @normanmaurer @trustin When you have a moment, could you please review the PR? There are several projects currently blocked, so your attention would be greatly appreciated. Thank you!

**Comment by altro3** [comment]
@yawkat @anna-geller This project is dead as you see. I think the single solution - fork this repo or create different and implement micronaut own module

**Comment by violetagg** [comment]
@anna-geller This project builds against `Netty 5` which we don't know when will be released.
The source code for `Netty 4.1/4.2` `codec-multipart` is located https://github.com/netty/netty/tree/4.2/codec-http/src/main/java/io/netty/handler/codec/http/multipart

**Comment by karelkryda** [comment]
@yawkat, would it be possible to send PR to the 4.2 branch as well?
@violetagg on your part, would you incorporate that PR into version 4.2 as well?

**Comment by yawkat** [comment]
> This PR is against netty 5, but if it goes well, I plan to backport the streaming parser to netty 4. My suggestion is to keep the backport in netty-contrib, release a netty 4 version of codec-multipart, and eventually deprecate the form decoder in the netty 4 core repo.

**Comment by karelkryda** [comment]
@violetagg, could you please give us your opinion? I think it is important to firmly state the procedure that will lead us to fix the parsing of multiform data.

You said that this repo is for netty 5. Do you agree that a corresponding repo would be created for netty 4, where @yawkat could send his PR?

If you agree with this, could you, @yawkat, possibly send the PR straight to the new repo and leave this PR for future netty 5 or do you insist that it must first be merged here?

I would like to avoid unnecessary misunderstandings from all parties involved.

Thanks in advance

**Comment by violetagg** [comment]
> > This PR is against netty 5, but if it goes well, I plan to backport the streaming parser to netty 4. My suggestion is to keep the backport in netty-contrib, release a netty 4 version of codec-multipart, and eventually deprecate the form decoder in the netty 4 core repo.

I just checked the PR, there are some files that IMO should not be there.
This version of the `codec-multipart` heavily depends on the new Netty 5 APIs so I doubt it will be easier to back port it to Netty 4.2.x version. May be you can create a branch with the sources from Netty 4.2.x and make the changes on top of them.

**Comment by yawkat** [comment]
I actually already worked on the 4.2 backport back when I made this PR: https://github.com/yawkat/codec-multipart/commits/multipart-refactor-4.2/

It is two commits behind, but the majority is done.

**Comment by anna-geller** [comment]
Hey 👋 just checking in, any updates on this?

**Comment by yuri1969** [comment]
@yawkat As I understand it, the `multipart-refactor-4.2` branch is currently missing some changes from the three newer commits in the original `multipart-refactor` branch - is that correct?

I've experimented with some of the changes in `multipart-refactor-4.2` against the current Netty 4.2, and it seems generally compatible, aside from source version compatibility issues, CheckStyle nagging, etc.

Do we have an idea what needs to be done to create a viable PR targeting Netty 4.2.x?

**Comment by yawkat** [comment]
There's a few of changes in the 5.0 branch that aren't in the 4.x one, I think the biggest one is lack of support for multipart/mixed. It'll take some effort to backport, but it's doable.

The main roadblock for this PR is that there are few people that are both interested and have the necessary netty knowledge to properly review it. The apple and spring folks could do it, but they're pretty busy with other stuff.

**Comment by yuri1969** [comment]
Working around the codec-multipart issue has become rather cumbersome/costly. Is there any way to raise the severity here?

What are the options? Would Netty sponsorship have any impact?

**Comment by karelkryda** [comment]
Hello,
I would like to lend a hand in this situation again.

It seems that you, @yawkat, still haven't created a PR for the **Netty 4** repository. I know you mentioned that the backport code for this version exists, but that there were some minor details left to finish.
So I would like to ask, is there a problem that has prevented you from creating a PR for the Netty 4 repo?
_If you think it's not worth the effort due to the poor prospects of merging into Netty 4, perhaps the following paragraph will resolve the issue._

@violetagg, I would like to ask you a question too.
I know you have a lot of work to do and it is difficult to devote time to such a large PR, but would it be possible for you to _"promise"_ in some way that after opening the PR for the **Netty 4** repo, there will be a code review and successful merging within a reasonable time? I know that "within a reasonable time" is a very broad term, but as has already been said, constantly dealing with this problem in a production environment costs us a lot of effort and money.

From my point of view, this problem affects a large number of people who use Netty - and other frameworks that use Netty.
I believe that together we can finally solve this problem once and for all.

**TLDR:** I would like to ask all parties involved to say what needs to be done to move things forward.

Thank you all very much in advance.

**Comment by yawkat** [comment]
@karelkryda norman has agreed that I can move forward on my own in this repo, since there seems to be lack of people that are both willing to and capable of reviewing my changes. My plan is to make a 4.2-based release for people to test. Due to the amount of work left to be done, this may still take a while though – this is only medium priority for me.

**Comment by karelkryda** [comment]
Thanks for the information, @yawkat, that sounds promising.
As @yuri1969 already mentioned, if a financial contribution would help, we are willing to discuss it.

We would also appreciate it if you could keep us informed about the progress 🙂.

Thanks again.

**Comment by violetagg** [comment]
I also agree that the mode `commit then review` is more appropriate for this project.

**Comment by yawkat** [comment]
merged manually

**Comment by yawkat** [comment]
@loicmathieu @anna-geller would you be interested in trying this? I've just released [Micronaut Core 4.10.9](https://github.com/micronaut-projects/micronaut-core/releases/tag/v4.10.9) which adds support for the fixed API. If you add [netty-codec-multipart-vintage 1.0.0.Alpha1](https://mvnrepository.com/artifact/io.netty.contrib/netty-codec-multipart-vintage/1.0.0.Alpha1) to the classpath, it should pick up the new decoder with some of the bugs you see fixed.

**Comment by loicmathieu** [comment]
@yawkat yes, we would be interested to test it.
However our user that reports the issue is not on our latest version so I'm not sure he can migrate to Micronaut Core 4.10.9, I'll have a look and keep you posted.
Thanks a lot for working on this.

**Comment by yawkat** [comment]
theres nothing that changed prior to 4.10.9, so i have a hard time believing that. maybe the issue is just too flaky for them to notice.

anyway, thanks, please keep me posted.

**Comment by loicmathieu** [comment]
@yawkat what I mean is that the issue only occurs very rarely and the user that reported it is on an old version of Kestra which I'm not sure I could migrate to the new version of Micronaut Core. Anyway, I'll see if it's possible or not to backport this change down to the version the user is using.

**Comment by loicmathieu** [comment]
@yawkat we're still using Micronaut 4.9 as 4.10 didn't work for us so we may not be able to test it very soon.
I'll see why Micronaut 4.10.2 didn't fix the issue we had with 4.10.0, I'll add feedback on the issue on Micronaut 4.10 migration.

**Comment by yuri1969** [comment]
Xref to the Micronaut 4.10.x issue: https://github.com/micronaut-projects/micronaut-core/issues/12157

**Comment by yawkat** [comment]
You may have to manually upgrade micronaut-micrometer.

**Comment by loicmathieu** [comment]
@yawkat I tried the provided netty-codec-multipart-vintage lib but it fails with some multipart upload, more specifically with ZIP upload with an error `java.util.zip.ZipException: invalid distance too far back`.

**Comment by yawkat** [comment]
could you provide a pcap of the input, if it's not confidential?

**Comment by loicmathieu** [comment]
We have an export/import feature which export as ZIP and import.
We have a test that know fails with the exception:

```java
    @Test
    void importFlowsWithZip() throws IOException {
        // create a ZIP file using the extract endpoint
        byte[] zip = client.toBlocking().retrieve(HttpRequest.GET("/api/v1/main/flows/export/by-query?namespace=io.kestra.tests"),
            Argument.of(byte[].class));
        File temp = File.createTempFile("flows", ".zip");
        Files.write(temp.toPath(), zip);

        var body = MultipartBody.builder()
            .addPart("fileUpload", "flows.zip", temp)
            .build();
        var response = client.toBlocking().exchange(POST("/api/v1/main/flows/import", body).contentType(MediaType.MULTIPART_FORM_DATA));

        assertThat(response.getStatus().getCode()).isEqualTo(OK.getCode());
        temp.delete();
    }
```

Here is the pcap extract, apparently the endpoint returns a status code 400.
[multipart-issue.zip](https://github.com/user-attachments/files/24523936/multipart-issue.zip)
